DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,337 @@
+    using DigitalLearningSolutions.Data.Models.Register;
+    using DigitalLearningSolutions.Data.Models.User;
+    using Microsoft.AspNetCore.Http;
+
+    public class BulkUploadResult
+    {
+        public enum ErrorReasons
+        {
+            InvalidJobGroupId,
+            InvalidLastName,
+            InvalidFirstName,
+            InvalidActive,
+            NoRecordForDelegateId,
+            UnexpectedErrorForUpdate,
+            UnexpectedErrorForCreate,
+            ParameterError,
+            AliasIdInUse,
+            EmailAddressInUse
+        }
+
+        public BulkUploadResult(
+            int processed,
+            int registered,
+            int updated,
+            int skipped,
+            IEnumerable<(int RowNumber, ErrorReasons Reason)> errors
+        )
+        {
+            Processed = processed;
+            Registered = registered;
+            Updated = updated;
+            Skipped = skipped;
+            Errors = errors;
+        }
+
+        public IEnumerable<(int RowNumber, ErrorReasons Reason)> Errors { get; set; }
+        public int Processed { get; set; }
+        public int Registered { get; set; }
+        public int Updated { get; set; }
+        public int Skipped { get; set; }
+    }
+
+    public interface IDelegateUploadFileService
+    {
+        public BulkUploadResult ProcessDelegatesFile(IXLTable table, int centreId, DateTime? welcomeEmailDate = null);
+        public IXLTable OpenDelegatesTable(IFormFile file);
+    }
+
+    public class DelegateUploadFileService : IDelegateUploadFileService
+    {
+        private const string DelegatesSheetName = "DelegatesBulkUpload";
+        private readonly IJobGroupsDataService jobGroupsDataService;
+
+        private readonly IRegistrationDataService registrationDataService;
+        private readonly IUserDataService userDataService;
+
+        public DelegateUploadFileService(
+            IJobGroupsDataService jobGroupsDataService,
+            IUserDataService userDataService,
+            IRegistrationDataService registrationDataService
+        )
+        {
+            this.userDataService = userDataService;
+            this.registrationDataService = registrationDataService;
+            this.jobGroupsDataService = jobGroupsDataService;
+        }
+
+        public BulkUploadResult ProcessDelegatesFile(IXLTable table, int centreId, DateTime? welcomeEmailDate)
+        {
+            if (!ValidateHeaders(table))
+            {
+                throw new InvalidHeadersException();
+            }
+
+            var jobGroupIds = jobGroupsDataService.GetJobGroupsAlphabetical().Select(item => item.id).ToList();
+            var (registered, updated, skipped) = (0, 0, 0);
+            var errors = new List<(int, BulkUploadResult.ErrorReasons)>();
+            var delegateRows = table.Rows().Skip(1).Select(row => new DelegateTableRow(table, row));
+
+            foreach (var delegateRow in delegateRows)

```

Comment:
This foreach loop is very long, we need to break down its contents into private methods. Maybe validation, Update and Insert?

Reply:
I'm not sure how to do that while keeping track of processed/registered/updated counters and errors.

Reply:
You should be able to pass them as reference parameters, or do something having a return type that returns updated values of them all.